File: Java/binary-tree-maximum-path-sum.java
//  Binary Tree Maximum Path Sum

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// Time O(N), Space O(1)
class Solution {
    private int max_sum = Integer.MIN_VALUE;
    
    public int maxPathSum(TreeNode root) {
        dfs(root);
        return max_sum;
    }
    
    private int dfs(TreeNode root) {
        if (root == null)
            return 0;
        int l_sum = dfs(root.left);
        int r_sum = dfs(root.right);
        int sum = root.val;
        if (l_sum > 0)
            sum += l_sum;
        if (r_sum > 0)
            sum += r_sum;
        max_sum = Math.max(max_sum, sum);
        return Math.max(l_sum, r_sum)>0 ? Math.max(l_sum, r_sum) + root.val : root.val;
    }
}

File: Leetcode150/binary-tree-maximum-path-sum.cc
//  Binary Tree Maximum Path Sum

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int max_sum = INT_MIN;
        maxPathSum(root, max_sum);
        return max_sum;
    }
    
    int maxPathSum(TreeNode *root, int &max_sum) {
        if (!root)  return 0;
        int l = maxPathSum(root->left, max_sum);
        int r = maxPathSum(root->right, max_sum);
        int sum = root->val;
        if (l > 0)  sum += l;
        if (r > 0)  sum += r;
        max_sum = max(max_sum, sum);
        return max(l, r)>0 ? max(l, r) + root->val : root->val;
    }
};
